erasure: replace hash-based bitrot detection with generic mechanism

aead opened the pull request:
This change introduces the new pkg/bitrot package and include it into the minio server.

## Description
This change replace the old hash-based bitrot detection approach with a new (more generic) solution.
Now it is possible to use hash-functions or authenticators (esp. polynomial authenticators).
Therefore this change adds a new field in the XL format (checkSumInfo now has also a key).

## Motivation and Context
This is the first step to support server-side-encryption (SSE).
Updates #4416 

## How Has This Been Tested?
Add test to ensure that currently just hash-functions (or similar non-key based techniques) can be used.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
vendor: add poly1305 and polynomial interface

Add the polynomial authenticator poly1305 and a general interface
for poly. authenticators to the vendor directory.

Commit:
add new bitrot package

The bitrot package provides general types and functions for
bitrot detection algorithms.

Commit:
introduce the pkg/bitrot in the minio server

This change replaces the old bitrot code with the pkg/bitrot package.
It is the first step of moving from a hash function based bitrot detection
approach to an encryption & authentication based one.

Instead of using a hash.Hash interface for bitrot checksums it is now also
possible to use other detection techniques (like poly. authenticators).

Commit:
erasure: add additional test to detect invalid algorithms

Add a test to ensure that currently only bitrot algorithms without keys
can be used.

harshavardhana commented on pkg/bitrot/bitrot.go:
```diff
@@ -0,0 +1,122 @@
+package bitrot
```
License header needed.

aead replied:
Done

harshavardhana commented on cmd/xl-v1-object.go:
```diff
@@ -266,7 +267,7 @@ func (xl xlObjects) GetObject(bucket, object string, startOffset int64, length i
 		checkSums := make([]string, len(onlineDisks))
-		var ckSumAlgo HashAlgo
+		var ckSumAlgo bitrot.Algorithm = 0xffffffff
 		for index, disk := range onlineDisks {
```
Any reason why choosing `0xffff` to indicate a default value . Can we use something else ?

aead replied:
It is an invalid value (because it much to high) Currently valid is 
 - 0: Poly1305
 - 1: GHASH
 - 2: BLAKE2b512
 - 3: SHA-256

Your right, that should be changed - Maybe making [unknownAlgorithm](https://github.com/aead/minio/blob/erasure/pkg/bitrot/bitrot.go#L27) visible? Then it should be clear...

harshavardhana replied:
Yes it is better that way @aead

harshavardhana commented on cmd/erasure-utils_test.go:
```diff
@@ -158,3 +161,21 @@ func TestCopyBuffer(t *testing.T) {
 		}
 	}
 }
+
```
Add a comment on the non obvious that this functionality is testing.

aead replied:
Done

Commit:
erasure: add package header and test doc
add licence header for bitrot package and
document test for bitrot algorithms

fwessels commented on cmd/posix.go:
```diff
@@ -521,8 +521,7 @@ func (s *posix) ReadAll(volume, path string) (buf []byte, err error) {
 func (s *posix) ReadFile(volume, path string, offset int64, buf []byte) (n int64, err error) {
-
-	return s.ReadFileWithVerify(volume, path, offset, buf, "", "")
+	return s.ReadFileWithVerify(volume, path, offset, buf, 0xfffffff, "")
```
Why are we passing (ugly) constant 0xffffffff (obviously -1) in ?
Make a proper const definition?

aead replied:
Same as Harsha already said, good catch! Done.

fwessels commented on cmd/posix_test.go:
```diff
@@ -1062,44 +1064,44 @@ func TestPosixReadFileWithVerify(t *testing.T) {
 		// Empty hash-algo returns error - 7
 		{
-			"myobject", 7, 4, "", blakeHash("Hello, world!"),
+			"myobject", 7, 4, 0xffffffff, blakeHash("Hello, world!"),
```
Same here (and in line 1104)

aead replied:
Done

fwessels commented on pkg/bitrot/bitrot.go:
```diff
@@ -0,0 +1,141 @@
+// It returns an error if the given key cannot be used by the algorithm.
+// New panics if the algorithm is unknown.
+func (a Algorithm) New(key []byte) (Hash, error) {
+	if keysize := a.KeySize(); keysize != len(key) {
```
Why not simply do `if a.KeySize() != len(key) {` etc.

aead replied:
Done

fwessels commented on pkg/bitrot/bitrot.go:
```diff
@@ -0,0 +1,141 @@
+// KeySize returns the size of valid keys for the given algorithm in bytes.
+// It panics if the algorithm is unknown.
+func (a Algorithm) KeySize() int {
+	if a >= 0 && a < UnknownAlgorithm {
```
a is a `uint` so `a >= 0` is a NOP.
Also you can better test against len(keysizes) -- in case somebody would add it to the const array but forget to add it to the keysizes array.

aead replied:
Done

fwessels commented on pkg/bitrot/bitrot.go:
```diff
@@ -0,0 +1,141 @@
+// String returns a string representation of the algorithm.
+// It panics if the algorithm is unknown.
+func (a Algorithm) String() string {
+	if a >= 0 && a < UnknownAlgorithm {
```
Idem

aead replied:
Done

fwessels commented on pkg/bitrot/bitrot.go:
```diff
@@ -0,0 +1,141 @@
+			return Algorithm(i), nil
+		}
+	}
+	return UnknownAlgorithm, fmt.Errorf("bitrot: algorithm #%s is unknown", s)
```
`#%s` --> `%s` as s is not a number

aead replied:
Done

fwessels commented on vendor/github.com/aead/poly/poly.go:
```diff
@@ -0,0 +1,28 @@
+package poly
```
May consider adding license header

aead replied:
Oh, yes, thank you!

fwessels commented on vendor/github.com/aead/poly1305/poly1305_amd64.go:
```diff
@@ -0,0 +1,108 @@
+		p.off = 0
+	}
+
+	if nn := len(msg) & (^(TagSize - 1)); nn > 0 {
```
Add comment why you are rounding up to multiple of TagSize here?

aead replied:
Maybe a good idea :wink:

fwessels commented on vendor/github.com/aead/poly1305/poly1305_amd64.go:
```diff
@@ -0,0 +1,108 @@
+	return n, nil
+}
+
+// Sum appends the Pol1305 hash of the previously
```
Pol1305 --> Poly1305

aead replied:
Done

fwessels commented on vendor/github.com/aead/poly1305/poly1305_amd64.go:
```diff
@@ -0,0 +1,108 @@
+	state := p.state
+
+	if p.off > 0 {
+		update(&state, p.buf[:p.off])
```
reset p.off to 0 here?

aead replied:
No, that would break the commented guarantee:
> // It is safe to call this function multiple times.

If we set `p.off = 0` the second call to Sum would return a different hash. That's the reason why we copy the state (see> `state := p.state`)

fwessels replied:
Ok, clear.

fwessels commented on vendor/github.com/aead/poly1305/poly1305_amd64.s:
```diff
@@ -0,0 +1,48 @@
+	MOVQ key+8(FP), SI
+
+
+	// state[0...7] is initialized with zero
```
This comment is not correct -- or do you mean: state[0...7] has been initialized with zero
(previously)

aead replied:
Yes, I mean that it is initialized with zeros when `initialize` is invoked (golang memory rule).
Otherwise a `PXOR X0, X0` and `PXOR X1, X1` would be necessary.

fwessels replied:
Ok

fwessels commented on vendor/github.com/aead/poly1305/poly1305_ref.go:
```diff
@@ -0,0 +1,228 @@
+		msg = msg[n:]
+	}
+	if len(msg) > 0 {
+		off += copy(block[:], msg)
```
`off +=` is equal to `off =` ?

aead replied:
Yes, it should... off is zero - I change this to `off = copy...`

fwessels commented on vendor/github.com/aead/poly1305/poly1305_ref.go:
```diff
@@ -0,0 +1,228 @@
+	var (
+		h, r  [5]uint32
+		s     [4]uint32
+		block [TagSize]byte
```
Move `block` and `off` inside of `if` statement in line 35?

aead replied:
Can be done, is this better? Does the compiler not alloc the memory if it is bound to the if scope?

fwessels replied:
It is not necessarily better (compiler-wise), maybe a little neater to allocate more closely to where you need it (and bound to scope).

fwessels commented on vendor/github.com/aead/poly1305/poly1305_ref.go:
```diff
@@ -0,0 +1,228 @@
+
+// Hash implements a Poly1305 writer interface.
+// Poly1305 cannot be used like common hash.Hash implementations,
+// because of using a poly1305 key twice breaks its security.
```
because of using --> because using

aead replied:
Done, thanks.

fwessels commented on vendor/github.com/aead/poly1305/poly1305_ref.go:
```diff
@@ -0,0 +1,228 @@
+		copy(buf[:], p.buf[:p.off])
+		buf[p.off] = 1 // invariant: p.off < TagSize
+
+		update(buf[:], finalBlock, &h, &(p.r))
```
reset p.off to 0 ?

aead replied:
Same here: // It is safe to call this function multiple times.

fwessels commented on vendor/github.com/aead/poly1305/poly1305_go17_amd64.s:
```diff
@@ -0,0 +1,197 @@
+
+// func supportsBMI2() bool
+TEXT ·supportsBMI2(SB), 4, $0-1
+    MOVQ runtime·support_avx2(SB), BX
```
Formatting

aead replied:
Done

Commit:
update bitrot (vendor) package
- fix ugly constant usage
- update vendord poly and poly1305 packages to last revision

aead commented on cmd/xl-v1-object.go:
```diff
@@ -266,7 +267,7 @@ func (xl xlObjects) GetObject(bucket, object string, startOffset int64, length i
 		// Get the checksums of the current part.
 		checkSums := make([]string, len(onlineDisks))
-		var ckSumAlgo HashAlgo
+		var ckSumAlgo bitrot.Algorithm = bitrot.UnknownAlgorithm
```
Done

fwessels commented on pkg/bitrot/bitrot.go:
```diff
@@ -0,0 +1,141 @@
+// String returns a string representation of the algorithm.
+// It panics if the algorithm is unknown.
+func (a Algorithm) String() string {
+	if a < UnknownAlgorithm && int(a) < len(keysizes) {
```
this should be `len(names)`

Commit:
pkg/bitrot: fix String() check
update vendor/github.com/aead/poly1305 to last revision

Commit:
pkg/bitrot: add AEAD cipher constants
Add AES-GCM and ChaCha20poly1305 AEADs as bitrot algorithms.
Those algorithms can be used to encrypt the data while still detect
data modification.

Commit:
pkg/bitrot: add Mode type

This change adds a Mode type. The mode is used to decide whether we
want to protect or verify data.
This is necessary because the encryption (protection) is
different from the decryption (verification). So we must somehow specify
for which purpose we want to use the cipher.

fwessels commented on pkg/bitrot/bitrot.go:
```diff
@@ -23,6 +23,16 @@ import (
+const (
+	// Protect indicates that the Hash is used for computing a checksum.
+	Protect Mode = true
+	// Verify indicates that the Hash is used for verifying a previous computed checksum.
```
previous --> previously

aead replied:
Done

fwessels commented on vendor/github.com/aead/poly1305/poly1305_ref.go:
```diff
@@ -34,7 +32,8 @@ func Sum(msg []byte, key [32]byte) [TagSize]byte {
 	}
 	if len(msg) > 0 {
-		off = copy(block[:], msg)
+		var block [TagSize]byte
```
If len(msg) is a multiple of 16 bytes (and this the remainder = 0, there is no update() using the finalBlock?
Is that correct behaviour?

aead replied:
Yes, this is intended behavior.

fwessels commented on vendor/github.com/aead/poly1305/poly1305_ref.go:
```diff
@@ -34,7 +32,8 @@ func Sum(msg []byte, key [32]byte) [TagSize]byte {
 	if len(msg) > 0 {
-		off = copy(block[:], msg)
+		var block [TagSize]byte
+		off := copy(block[:], msg)
```
if len(msg) ever is TagSize, then the following statement `block[off] = 1` would cause a buffer overrun -- may want to protect against this?

aead replied:
No, I don't think so: If `len(msg) == TagSize` than `len(msg) & (^(TagSize - 1)) == 16` 
I also ran:
```
func TestOverflow(t *testing.T) {
	msg := make([]byte, TagSize)
	var key [32]byte
	Sum(msg, key)
}
with 'export GOARCH=386' and 'go run -v -run=TestOverflow'

Commit:
pkg/bitrot: fix typo
